query-engine/core/src/query_document/parser.rs

```diff
@@ -252,14 +258,33 @@ impl QueryDocumentParser {
             }
             (QueryValue::DateTime(s), ScalarType::DateTime) => Ok(PrismaValue::DateTime(s)),
 
-            (QueryValue::Int(i), ScalarType::Int) => Ok(PrismaValue::Int(i)),
+            (QueryValue::Int(i), ScalarType::Int(int_type)) => {
+                let fits_in_specified_type = match &int_type {
+                    IntType::Signed8 => i8::try_from(i).is_ok(),
+                    IntType::Signed16 => i16::try_from(i).is_ok(),
+                    IntType::Signed24 => fits_in_range(i, I24_MIN, I24_MAX),
+                    IntType::Signed32 => i32::try_from(i).is_ok(),
+                    IntType::Unsigned8 => u8::try_from(i).is_ok(),
+                    IntType::Unsigned16 => u16::try_from(i).is_ok(),
+                    IntType::Unsigned24 => fits_in_range(i, U24_MIN, U24_MAX),
+                    IntType::Unsigned32 => u32::try_from(i).is_ok(),
+                    IntType::Custom(min, max) => fits_in_range(i, *min, *max)
+                };
+
+                if fits_in_specified_type {
+                    Ok(PrismaValue::Int(i))
+                } else {
+                    Err(QueryParserError::new(parent_path.clone(), QueryParserErrorKind::ValueFitError(
+                        format!("Unable to fit integer value '{}' into a {} for field '{}'.", i, int_type, parent_path.last().unwrap()))))
+                }
+            },
             (QueryValue::Int(i), ScalarType::Float) => Ok(PrismaValue::Float(BigDecimal::from(i))),
             (QueryValue::Int(i), ScalarType::Decimal) => Ok(PrismaValue::Float(BigDecimal::from(i))),
             (QueryValue::Int(i), ScalarType::BigInt) => Ok(PrismaValue::BigInt(i)),
 
             (QueryValue::Float(f), ScalarType::Float) => Ok(PrismaValue::Float(f)),
             (QueryValue::Float(f), ScalarType::Decimal) => Ok(PrismaValue::Float(f)),
-            (QueryValue::Float(f), ScalarType::Int) => match f.to_i64() {
+            (QueryValue::Float(f), ScalarType::Int(_)) => match f.to_i64() {
                 Some(converted) => Ok(PrismaValue::Int(converted)),
```

Comment:
I don't know if there's a good place to persist that, but it would be good to document that we validate all integers except 64 bit integers because they are strings at the graphql level (IIUC).